crates/sc-proof-of-time/src/time_keeper.rs

```diff
@@ -43,55 +42,32 @@ where
     Client: HeaderBackend<Block> + BlockchainEvents<Block>,
 {
     /// Creates the time keeper instance.
-    pub fn new<Network, GossipSync>(
-        components: PotComponents,
+    pub fn new(
+        components: &PotComponents,
         client: Arc<Client>,
-        network: Network,
-        sync: Arc<GossipSync>,
         target_proof_time: Duration,
-    ) -> Self
-    where
-        Network: GossipNetwork<Block> + Send + Sync + Clone + 'static,
-        GossipSync: GossipSyncing<Block> + 'static,
-    {
-        let PotComponents {
-            proof_of_time,
-            protocol_state: pot_state,
-            ..
-        } = components;
-
+        gossip_sender: mpsc::Sender<PotProof>,
+    ) -> Self {
         Self {
-            proof_of_time: proof_of_time.clone(),
-            pot_state: pot_state.clone(),
-            gossip: PotGossip::new(network, sync, pot_state, proof_of_time),
+            proof_of_time: components.proof_of_time,
+            pot_state: Arc::clone(&components.protocol_state),
             client,
             target_proof_time,
+            gossip_sender,
+            _block: PhantomData,
         }
     }
 
     /// Runs the time keeper processing loop.
-    pub async fn run(self) {
+    pub async fn run(mut self) {
```

Comment:
We could potentially get rid of run() completely?, as it only proxies messages to gossip worker after this change

The producer thread can use something like https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on to directly send message to gossip worker

Reply:
`block_on` is a blocking operation, not sure how you want to use it here. `fn run` returns future that is driven by something external. If something external doesn't need it anymore (node is shutting down for instance), then future will be dropped.

Reply:
The producer thread is a std thread, does this currently:
```

            } else if let Err(error) = proof_sender.blocking_send(next_proof.clone()) {
                warn!(%error, "Produce proofs: send failed");
                return;
            }
```

I don't understand your concern

Reply:
`blocking_send` is happening from a dedicated thread, `run` is an async function. I don't understand what you're proposing here.

Reply:
I temporarily abandoned Timekeeper as a data structure in the local WIP branch, didn't see the point in a struct, will have to see what to do about it once I get to connecting it with gossip again.